# Another play on `words`
:face_with_rolling_eyes:

## Objectives
1. Explore basic Python capabilities for a known dataset
2. Explore file input/output, string manipulation, loops and basic constructs in Python

## Instructions
1. Repeat portions of last week's `words` 01_shell_exercise using Python and the Jupyter notebook
2. Turn in the completed Jupyter notebook containing code and output

## 0. Read the file and create a list of strings for subsequent analysis
*Note: you will need to handle newline strings '\n' at the end of each word*

In [1]:
wordList = open('words').read().splitlines()

## 1. How many words are there in the file?

In [2]:
len(wordList)

235886

## 2. How many characters total?
*Use list comprehension*

In [5]:
# This is different from what the shell work found.
# read into this and it seems like python does different stuff
# with special characters and encoding??
# I don't think this should be an issue with this list of words though...
sum([len(word) for word in wordList])

2257223

## 3. Print the first 3 words, print the last 3 words
*Use relative indices for list slicing (i.e., don't use your length from question #1)*

In [6]:
print(wordList[:3])
print(wordList[-3:])

['A', 'a', 'aa']
['zythum', 'Zyzomys', 'Zyzzogeton']


## 4. How many characters are in the longest word?

In [7]:
max([len(word) for word in wordList])

24

## 5. Does your list contain the nickname for the UW mascot?
*This should be simple boolean statement*

In [8]:
'dubs' in wordList

True

## 6. If so, what is the index for that word?  Check by printing the word at that index.

In [9]:
print(wordList.index('dubs'))
wordList[58209]

58209


'dubs'

## 7. Create a function to accomplish #5 and #6
* Function should accept two arguments, the list of words and the word you want to find
* If the word is found, the function should return the index for that word, otherwise return `None`

In [10]:
def findWord(lister,word):
    try:
        return lister.index(word)
    except:
        return None
findWord(wordList,'dubs')        

58209

## 8. How many words begin with each letter of the alphabet (case-insensitive)?
Hints:
* Python has built-in list of lowercase letters stored as string.ascii_lowercase
* There are a few ways to approach this, but most straightforward is probably to use nested loops:
    * Loop through each letter
        * Initialize some count variable or empty list
        * Loop through each word
            * Check to see if the word starts with the letter (careful about case!)
            * If it does, increment your counter or append the word to the list
        * Print out the letter and the total number of words            

In [11]:
import string
Lets = string.ascii_lowercase

for i in Lets:
    count = 0
    for word in wordList:
        if word[0].lower() == i:
            count += 1
    print(i,count)

a 17096
b 11070
c 19901
d 10896
e 8736
f 6860
g 6861
h 9027
i 8799
j 1642
k 2281
l 6284
m 12616
n 6780
o 7849
p 24461
q 1152
r 9671
s 25162
t 12966
u 16387
v 3440
w 3944
x 385
y 671
z 949


## 99. Extra Credit (or if you finish the above exercises early)
Create a standalone python script that can be executed on the command line:
* Use sys.argv to accept command line argument for the `words` filename location
* Print the results to stdout
* Save the results to a file called `words_lettercount.txt`
    * Use the os.path module to append the '_lettercount.txt' string to the original filename
    * Output file should include a letter and total count on each line (“a 17096”)